tests/Unit/DataStructures/VectorImplTestHelper.hpp

```diff
@@ -2,17 +2,77 @@
 // See LICENSE.txt for details.
 #pragma once
 
+#include <array>
 #include <cstddef>
 #include <memory>  // IWYU pragma: keep
 #include <tuple>
 #include <utility>
 
+#include "DataStructures/VectorImpl.hpp"
 #include "Utilities/Gsl.hpp"
 #include "Utilities/Requires.hpp"
 #include "Utilities/Tuple.hpp"
+#include "tests/Unit/TestingFramework.hpp"
 
 namespace TestHelpers {
 namespace VectorImpl {
+
+namespace detail {
+// `check_vectors` checks equality of vectors and related data types for tests.
+//
+// Comparisons between any pairs of the following are supported:
+// vector types
+// arithmetic types
+// arrays of vectors
+// arrays of arithmetic types
+
+// between two vectors
+template <typename T1, typename T2, typename VT1, typename VT2>
+inline void check_vectors(const ::VectorImpl<T1, VT1>& t1,
```

Comment:
Why are these in a detail namespace?

Reply:
@markscheel felt their behavior for comparing against contained types was peculiar enough to avoid them as public interfaces.

Reply:
Looking at this more closely, I'm confused about what this file is for (which makes it difficult to judge what its public interface should be).  From the name, I believe it to be helpers for testing VectorImpl, but it isn't included in the test of VectorImpl and does not appear to be used anywhere except in the test of itself.  I assume some consumers will be added in a future pull request.

If those consumers are not going to use these functions, then they can be moved into Test_VectorImplTestHelper.cpp.  If the other comsumers are expected to use these then it seems to me that that makes them part of the public interface.

If a public function behaves oddly it should be documented clearly.  Moving it to a detail namespace is the opposite of that, since it hides the documentation entirely.

Reply:
The plan (which should become clear in the next PR in the line) is to have a set of helper functions very similar to the existing `DataVector` tests, which will consume the `check_vectors`. The hope is that the actual tests of derived classes of `VectorImpl` should only have to use those intermediate helper functions and never have to call `check_vectors` directly.

If that's still inappropriate for the `detail`, I can move it back out and document the behavior more carefully.

Reply:
OK, that sounds reasonable.

I think perhaps this set of changes has been split into too many PRs, since I'm really struggling to see the big picture.